contracts/StrategyCurveTwoPool.sol

```diff
@@ -0,0 +1,372 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity 0.6.12;
+pragma experimental ABIEncoderV2;
+
+// These are the core Yearn libraries
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/math/Math.sol";
+
+import { ICurveGauge, ICurvePool } from "../interfaces/ICurveTwoPool.sol";
+import { ISwapRouter } from "../interfaces/ISwapRouter.sol";
+import { BaseStrategy, StrategyParams } from "@yearnvaults/contracts/BaseStrategy.sol";
+
+abstract contract StrategyCurveBase is BaseStrategy {
+  using Address for address;
+
+  /* ========== STATE VARIABLES ========== */
+  // these should stay the same across different wants (`2CRV` LP tokens, `crv3crypto` LP tokens, etc.).
+
+  ICurveGauge public immutable gauge;
+
+  ICurvePool public immutable curve;
+
+  address internal immutable wethAddress;
+
+  // Uniswap V3 router
+  ISwapRouter public immutable router;
+
+  IERC20 public crv;
+
+  bool internal forceHarvestTriggerOnce; // only set this to true externally when we want to trigger our keepers to harvest for us
+  uint256 public creditThreshold; // amount of credit in underlying tokens that will automatically trigger a harvest
+
+  string internal stratName; // set our strategy name here
+
+  /* ========== CONSTRUCTOR ========== */
+
+  constructor(
+    address _vaultAddress,
+    address _healthCheckAddress,
+    address _gaugeAddress,
+    address _poolAddress,
+    address _wethAddress,
+    address _crvAddress,
+    address _routerAddress
+  ) public BaseStrategy(_vaultAddress) {
+    // Curve contracts
+    gauge = ICurveGauge(_gaugeAddress);
+    curve = ICurvePool(_poolAddress);
+    crv = IERC20(_crvAddress);
+
+    // Uniswap V3 router
+    router = ISwapRouter(_routerAddress);
+
+    // Wrapped ETH - Used for multi-hop swap. No contract instance needed
+    wethAddress = _wethAddress;
```

Comment:
shouldn't we hardcode this?

Reply:
wethAddress is an `immutable` variable which is similar to `contant` from a gas standpoint. https://docs.soliditylang.org/en/v0.8.11/contracts.html#constant-and-immutable-state-variables